include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -344,9 +619,406 @@ class alignment_file_format_sam
+                                          " was given."};
+                }
+                break;
+            }
+            default:
+                throw parse_error{std::string("[SAM Parse Error] The second character in the numerical id of a "
+                                  "SAM tag must be one of [A,i,Z,H,B,f] but ") + type_id + "was given."};
+        }
+    }
+
+    /*!\brief Reads the SAM header_ptr.
+     * \tparam stream_view_type     The type of the stream as a view.
+     * \param[in, out] stream_view  The stream view to iterate over.
+     * \param[in, out] header_ptr   The header_ptr (as a pointer) to store the parsed values.
+     *
+     * \throws seqan3::parse_error if any unexpected character or format is encountered.
+     *
+     * \details
+     *
+     * Reading the header format is done according to the official
+     * [SAM format specifications](https://samtools.github.io/hts-specs/SAMv1.pdf).
+     *
+     * The function throws a seqan3::parse_error if any unknown tag was encountered. It will also fail if the format is
+     * not in a correct state (e.g. required fields are not given), but throwing might occur downstream of the actual
+     * error.
+     */
+    template <typename stream_view_type>
+    void read_header(stream_view_type && stream_view, std::unique_ptr<alignment_file_header> & header_ptr)
+    {
+        alignment_file_header & hdr{*header_ptr};
+
+        auto parse_tag_value = [&stream_view, this] (auto & value) // helper function to parse the next tag value
+            {
+                detail::consume(stream_view | view::take_until_or_throw(is_char<':'>)); // skip tag name
+                std::ranges::next(std::ranges::begin(stream_view));                     // skip ':'
+                read_field(stream_view | view::take_until_or_throw(is_char<'\t'> || is_char<'\n'>), value);
+            };
+
+        // @HQ line
+        // -------------------------------------------------------------------------------------------------------------
+        parse_tag_value(hdr.format_version); // parse required VN (version) tag
+
+        // The SO, SS and GO tag are optional and can appear in any order
+        while (is_char<'\t'>(*std::ranges::begin(stream_view)))
+        {
+            std::ranges::next(std::ranges::begin(stream_view));              // skip tab
+            std::string * who = std::addressof(hdr.grouping);
+
+            if (is_char<'S'>(*std::ranges::begin(stream_view)))
+            {
+                std::ranges::next(std::ranges::begin(stream_view));          // skip S
+
+                if (is_char<'O'>(*std::ranges::begin(stream_view)))          // SO (sorting) tag
+                    who = std::addressof(hdr.sorting);
+                else if (is_char<'S'>(*std::ranges::begin(stream_view)))     // SS (sub-order) tag
+                    who = std::addressof(hdr.subsorting);
```

Comment:
![grafik](https://user-images.githubusercontent.com/7955878/54343182-8e403500-463e-11e9-8495-1e483ae4770e.png)

ask @rrahn

Reply:
Then please tell me, Mr Rahn!

std::addressof is not a universal replacement for `&`. The `&` symbol is the normal way of getting the address. Some very few types overload this, like `semiregular_t<>` – and also with good reason!
Now, if you work with these specific types and you absolutely need the location of the shim class and not the contained object, you use std::addressof.

Using it in this place signals to the reader, that in fact `hdr.sorting` is something like `semiregular_t<std::string>` and that you want the behaviour above, i.e. it implies certain more complex semantics than `&` although they aren't there.

Reply:
I am pleased to tell you, Mr Hauswedell!

As you correctly observed std::addressof is not an universal replacement of the unary-`&` operator, which can be overloaded for user specific types.  But on the contrary, a lot of types you would expect they overload it, do in fact not - and also for a very good reason! 

Here are some examples that do not overload the &-operator, also it could make sense to do it.
```cpp
    int v = 0;
    foo f{};
    ranges::semiregular_t<foo> semi{f};
    ranges::semiregular_t<int &> semi_ref{v};
    std::unique_ptr<int> uptr{new int(4)};
    std::shared_ptr<int> sptr{new int(4)};
    std::reference_wrapper<int> ref{v};

    assert(std::addressof(semi) == &semi);
    assert(std::addressof(semi_ref) == &semi_ref);
    assert(std::addressof(uptr) == &uptr);
    assert(std::addressof(sptr) == &sptr);
    assert(std::addressof(ref) == &ref);
```

The [cpp core guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c166-overload-unary--only-as-part-of-a-system-of-smart-pointers-and-references) suggest to only use this in a system with proper smart pointers. Which is a very good thing to do, but something that can miserably fail if not done right. And I would assume, that this is something you can shoot yourself in the foot with very easily. That's why a major part of the c++ community discourages to do it, in fact there are c++ guidelines for medical devices and aviation software that forbid doing this.
Now of course, the header type is something that is unlikely to be overloaded or adapted by the user and we can be rather sure, that &-operator does what it should. Also, one never knows what future developers of the library are going to do, but still the use case seems very unlikely. 

>Using it in this place signals to the reader, that in fact hdr.sorting is something like semiregular_t<std::string> and that you want the behaviour above, i.e. it implies certain more complex semantics than & although they aren't there.

This is not right IMOP. Because it is very expressive in what it does. It does return the address of the respective object. I do know what I get. When using &, I am never sure  what I get, also I think I do.

**BUT**: When working with user types in generic code `std::addressof` is the only safe way. Otherwise, if you expect your code can handle types that do some tricky part you should explicitly write code for it. 
For me it is debatable if we should enforce this for types that are only used internally by the library, coming from the library. So with this regard I am more in favor of using std::addressof explicitly (since I believe strange errors are easier to detect then a misbehvaing &-operator), but I would not enforce it if there are valid reasons against it. You sure will find some 😉

Reply:
> The cpp core guideline suggest to only use this in a system with proper smart pointers. Which is a very good thing to do, but something that can miserably fail if not done right. And I would assume, that this is something you can shoot yourself in the foot with very easily. That's why a major part of the c++ community discourages to do it, in fact there are c++ guidelines for medical devices and aviation software that forbid doing this.
>Now of course, the header type is something that is unlikely to be overloaded or adapted by the user and we can be rather sure, that &-operator does what it should. Also, one never knows what future developers of the library are going to do, but still the use case seems very unlikely.

I fully agree with all of this and would discourage overloading `operator&` for almost all cases!

> BUT: When working with user types in generic code std::addressof is the only safe way. 

I disagree about this. I think overloading `operator&` is very, very niche. 99% of C++ developers have probably never heard about it. If one of the 1% decides that they want to overload it, then I assume they need this behaviour for the correct operation of their type (there must be some reason after all to do it) and we should probably not work around it. One reason could be that the object passed is a weird shim class around the actual content and that the shim class could go end-of-life before the content, and that therefore operator& forwards to the actual content. In that case our code would break, because we did not trust the intentions of the developer who overloaded the operator.

Could this cause problems for types with incorrectly specified `operator&` ? Yes, it could, but so could many other things, like syntactically modelling a concept, but semantically violating its requirements. And we shouldn't kill-off the valid use-cases to prevent invalid use-cases, especially for a feature that I assume only experts will use.

Reply:
> One reason could be that the object passed is a weird shim class around the actual content and that the shim class could go end-of-life before the content, and that therefore operator& forwards to the actual content. In that case our code would break, because we did not trust the intentions of the developer who overloaded the operator.

On the contrary, if my algorithm/data structure holds on to a value that is not the one passed to it, then I am in real trouble. Because then I have no guarantee that the pointed-to object wasn't modified or erased by other processes. If I get an address to the object that was passed to my generic code I will have it (not considering general pitfalls with dangling references and such). And of course I can make sure that the type passed fulfills the requirements so I can safely use it. Thus it is the responsibility of the developer (in this 1% of the cases) to make sure that it passes then the pointed-to object if this was the intention of the class design. It is not the responsibility of the algorithm data structure.